### PR TITLE
Recipe: Caches

### PR DESCRIPTION
**Pull request by Lotes**
I added a recipe about the document cache.
I am also mentioning the other caches (WorkspaceCache, ContextCache and SimpleCache).

I moved this recipe under the category "utilities". Maybe "performance" would be better?

### CODE REVIEW

File: hugo/content/docs/recipes/utilities/caches.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+        this.cache = new DocumentCache(services.shared);
+    }
+    override inferPublisher(person: Person): KnownPublisher|undefined {
+        const documentUri = AstUtils.getDocument(person).uri;
+        if(this.cache.has(documentUri, person)) {
+           return this.cache.get(documentUri, person)!;
+        }
+        const publisher = super.inferPublisher(person);
+        this.cache.set(documentUri, person, publisher);
+        return publisher;
```
Comment (lars-reimann) [inline hugo/content/docs/recipes/utilities/caches.md:84]: Demonstrating the use of `ContextCache.get(contextKey: Context, key: Key, provider: () => Value): Value` might also be useful:

```ts
const documentUri = AstUtils.getDocument(person).uri;
return this.cache.get(documentUri, person, () => super.inferPublisher(person));
```

Comment (Lotes) [reply hugo/content/docs/recipes/utilities/caches.md:84]: Uh! Very nice! I was not aware of, thank you!

File: hugo/content/docs/recipes/utilities/caches.md [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+## What is the problem?
+
+You have parsed a document and you would like to execute some computation on the AST. But you don’t want to do this every time you see a certain node. You want to do it once for the lifetime of a document. Where to save it?
+
+## How to solve it?
+
+For data that depends on the lifetime of a document or even the entire workspace, Langium has several kinds of caches.
+
+* the document cache saves key-value-pairs of given types `K` and `V` for each document. If the document gets changed or deleted the cache gets cleared automatically for the single files
+* the workspace cache also save key-value-pairs of given types `K` and `V` but gets cleared entirely when something in the workspace gets changed.
```
Comment (montymxb) [inline hugo/content/docs/recipes/utilities/caches.md:15]:
```suggestion
* the workspace cache also saves key-value-pairs of given types `K` and `V`, but gets cleared entirely when something in the workspace gets changed
```
small typo, and would drop the `.` to be consistent with the first bullet point

File: hugo/content/docs/recipes/utilities/caches.md [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+Besides those specific caches, Langium also provides
```
Comment (montymxb) [inline hugo/content/docs/recipes/utilities/caches.md:17]:
```suggestion
Besides those specific caches, Langium also provides:
```

File: hugo/content/docs/recipes/utilities/caches.md [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+weight: 0
+---
+
+## What is the problem?
+
+You have parsed a document and you would like to execute some computation on the AST. But you don’t want to do this every time you see a certain node. You want to do it once for the lifetime of a document. Where to save it?
+
+## How to solve it?
+
+For data that depends on the lifetime of a document or even the entire workspace, Langium has several kinds of caches.
```
Comment (montymxb) [inline hugo/content/docs/recipes/utilities/caches.md:12]:
```suggestion
For data that depends on the lifetime of a document or even the entire workspace, Langium has several kinds of caches:
```

File: hugo/content/docs/recipes/utilities/caches.md [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+* the document cache saves key-value-pairs of given types `K` and `V` for each document. If the document gets changed or deleted the cache gets cleared automatically for the single files
+* the workspace cache also save key-value-pairs of given types `K` and `V` but gets cleared entirely when something in the workspace gets changed.
+
+Besides those specific caches, Langium also provides
+
+* a simple cache that can be used for any kind of key-value-data
+* a context cache that stores a simple cache for each context object. The document cache and workspace cache are implemented using the context cache.
```
Comment (montymxb) [inline hugo/content/docs/recipes/utilities/caches.md:20]:
```suggestion
* a context cache that stores a simple cache for each context object. The document cache and workspace cache are implemented using the context cache
```

File: hugo/content/docs/recipes/utilities/caches.md [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+## How to use it?
+
+Here we will use the `HelloWorld` example from the learning section. Let's keep it simple and just list persons in the document, that come from a comic book.
```
Comment (montymxb) [inline hugo/content/docs/recipes/utilities/caches.md:24]:
```suggestion
Here we will use the `HelloWorld` example from the learning section. Let's keep it simple and just list people in a document, which will come from a comic book.
```

File: hugo/content/docs/recipes/utilities/caches.md [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+person Wonderwoman
+person Spiderman
+person Homer //warning: unknown publisher!!
+person Obelix
+```
+
+## Last words
+
+Caching can improve the performance of your language server. It is especially useful for computations that are expensive to calculate. The `DocumentCache` and `WorkspaceCache` are the most common caches to use. The `ContextCache` is useful if you need to store data for a specific context object. If you only need a key-value store, you can use the `SimpleCache`.
+All of them are disposable compared to a simple `Map<K, V>`. If you dispose them by calling `dispose()` the entries will be removed and the memory will be freed.
```
Comment (montymxb) [inline hugo/content/docs/recipes/utilities/caches.md:168]:
```suggestion
All of these caches are disposable compared to a simple `Map<K, V>`. If you dispose them by calling `dispose()` the entries will be removed and the memory will be freed.
```

**Review by montymxb — COMMENTED** [review]
Cool 🙂 . Just added some grammar suggestions, but the content looks really nice.

**Comment by Lotes** [comment]
@msujew Can you tell me whether there is a common way in Langium to register the disposal of a cache? I explained the method, but have not shown any connection points with Langium.

**Comment by msujew** [comment]
@Lotes Given that the language server process lives and dies with its connection (and therefore also service module), any global caches (i.e. caches that live as part of a service implementation) are "disposed" as soon as the language server dies. They don't need to be explicitly registered for disposal.

Making these caches disposable was designed for cases in which the cache itself is designed to be short-lived. I assume these cases are fairly rare (we have none in our framework) and there's no recommended way of performing the disposal for now.

File: hugo/content/docs/recipes/performance/caches.md [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+Here we will use the `HelloWorld` example from the learning section. Let's keep it simple and just list people in a document, which will come from a comic book.
+
+We will have a computation for each person that determines from which publisher it comes from.
+
+### Add a database
+
+Let's build a "publisher inferer service". First let's create a small database of known publishers and known persons:
+
+```typescript
+type KnownPublisher = 'DC'|'Marvel'|'Egmont';
```
Comment (msujew) [inline hugo/content/docs/recipes/performance/caches.md:33]: Formatting:
```suggestion
type KnownPublisher = 'DC' | 'Marvel' | 'Egmont';
```

File: hugo/content/docs/recipes/performance/caches.md [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+};
+```
+
+### Define the computation service
+
+For our service we define an interface:
+
+```typescript
+export interface InferPublisherService {
+    inferPublisher(person: Person): KnownPublisher|undefined;
```
Comment (msujew) [inline hugo/content/docs/recipes/performance/caches.md:47]: Formatting:
```suggestion
    inferPublisher(person: Person): KnownPublisher | undefined;
```

File: hugo/content/docs/recipes/performance/caches.md [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+class UncachedInferPublisherService implements InferPublisherService {
+    inferPublisher(person: Person): KnownPublisher|undefined {
+        for (const [publisher, persons] of Object.entries(KnownPersonNames)) {
+            if(persons.includes(person.name)) {
+                return publisher as KnownPublisher;
+            }
+        }
+        return undefined;
+    }
+}
```
Comment (msujew) [inline hugo/content/docs/recipes/performance/caches.md:63]: Formatting:
```suggestion
class UncachedInferPublisherService implements InferPublisherService {
    inferPublisher(person: Person): KnownPublisher | undefined {
        for (const [publisher, persons] of Object.entries(KnownPersonNames)) {
            if (persons.includes(person.name)) {
                return publisher as KnownPublisher;
            }
        }
        return undefined;
    }
}
```

File: hugo/content/docs/recipes/performance/caches.md [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        super();
+        this.cache = new DocumentCache(services.shared);
+    }
+    override inferPublisher(person: Person): KnownPublisher|undefined {
+        const documentUri = AstUtils.getDocument(person).uri;
+        //get cache entry for the documentUri and the person
+        //if it does not exist, calculate the value and store it
+        return this.cache.get(documentUri, person, () => super.inferPublisher(person));
+    }
+}
```
Comment (msujew) [inline hugo/content/docs/recipes/performance/caches.md:83]: Formatting:
```suggestion
export class CachedInferPublisherService extends UncachedInferPublisherService {
    private readonly cache: DocumentCache<Person, KnownPublisher | undefined>;
    constructor(services: HelloWorldServices) {
        super();
        this.cache = new DocumentCache(services.shared);
    }
    override inferPublisher(person: Person): KnownPublisher | undefined {
        const documentUri = AstUtils.getDocument(person).uri;
        //get cache entry for the documentUri and the person
        //if it does not exist, calculate the value and store it
        return this.cache.get(documentUri, person, () => super.inferPublisher(person));
    }
}
```

File: hugo/content/docs/recipes/performance/caches.md [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+
+## Last words
+
+Caching can improve the performance of your language server. It is especially useful for computations that are expensive to calculate. The `DocumentCache` and `WorkspaceCache` are the most common caches to use. The `ContextCache` is useful if you need to store data for a specific context object. If you only need a key-value store, you can use the `SimpleCache`.
+All of these caches are disposable compared to a simple `Map<K, V>`. If you dispose them by calling `dispose()` the entries will be removed and the memory will be freed.
+
+## Appendix
+
+<details>
+<summary>Ful implementation</summary>
```
Comment (msujew) [inline hugo/content/docs/recipes/performance/caches.md:173]: Typo. Also please fix the formatting in the full implementation.

```suggestion
<summary>Full Implementation</summary>
```

File: hugo/content/docs/recipes/performance/caches.md [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+ */
+export class HelloWorldValidator {
+    private readonly inferPublisherService: InferPublisherService;
+
+    constructor(services: HelloWorldServices) {
+        this.inferPublisherService = services.utilities.inferPublisherService;
+    }
+
+    checkPersonIsFromKnownPublisher(person: Person, accept: ValidationAcceptor): void {
+        if(this.inferPublisherService.inferPublisher(person) === undefined) {
```
Comment (msujew) [inline hugo/content/docs/recipes/performance/caches.md:119]: Formatting. Also comparisons with `undefined` are unusual in TS/JS. Just use the falsyness check
```suggestion
        if (!this.inferPublisherService.inferPublisher(person)) {
```

File: hugo/content/docs/recipes/performance/caches.md [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+person Wonderwoman
+person Spiderman
+person Homer //warning: unknown publisher!!
+person Obelix
+```
+
+## Last words
+
+Caching can improve the performance of your language server. It is especially useful for computations that are expensive to calculate. The `DocumentCache` and `WorkspaceCache` are the most common caches to use. The `ContextCache` is useful if you need to store data for a specific context object. If you only need a key-value store, you can use the `SimpleCache`.
+All of these caches are disposable compared to a simple `Map<K, V>`. If you dispose them by calling `dispose()` the entries will be removed and the memory will be freed.
```
Comment (msujew) [inline hugo/content/docs/recipes/performance/caches.md:168]: Suggestion: Normal maps have a `clear` method (and our caches do as well). The special thing about `dispose` in our caches is that they stop reacting to changes in the workspace once they are disposed. If they are not disposed, the reference to the cache event will stay in the `DocumentBuilder`,

Comment (Lotes) [reply hugo/content/docs/recipes/performance/caches.md]: Added.

**Comment by Lotes** [comment]
Thanks, I added your suggestions, @msujew .

**Review by msujew — APPROVED** [review]
Looks good to me, thanks 👍

**Pull request merged**